### ISSUE TITLE
Support ancestral state reconstruction

### ISSUE THREAD

**Issue by Anaphory** [lmaurits/BEASTling#128]
It would be nice to add ancestral state reconstruction to BEASTling.

**Comment by Anaphory** [comment]
Judging from the template in `beast-classic`, I assumed this was mostly swapping the `TreeLikelihood` with an `AncestralStateTreeLikelihood` and `TreeWithTraitLogger` instead of `TreeWithMetaDataLogger`. The other stuff in that template seems to mostly set up a BSVS model. It appears it's not that easy, or I'm committing some stupid mistakes again.

However, those swaps just give me
```
         Sample ESS(posterior)          prior     likelihood      posterior
Fatal exception: null
java.lang.NullPointerException
	at java.lang.System.arraycopy(Native Method)
	at beast.evolution.likelihood.BeerLikelihoodCore.getNodeStates(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.traverseSample(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.redrawAncestralStates(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood.getStatesForNode(Unknown Source)
	at beast.evolution.likelihood.AncestralStateTreeLikelihood$1.getTraitString(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.toNewick(Unknown Source)
	at beast.evolution.tree.TreeWithTraitLogger.log(Unknown Source)
	at beast.core.Logger.log(Unknown Source)
	at beast.core.MCMC.log(Unknown Source)
	at beast.core.MCMC.doLoop(Unknown Source)
	at beast.core.MCMC.run(Unknown Source)
	at beast.app.BeastMCMC.run(Unknown Source)
	at beast.app.beastapp.BeastMain.<init>(Unknown Source)
	at beast.app.beastapp.BeastMain.main(Unknown Source)
	at beast.app.beastapp.BeastLauncher.main(Unknown Source)
Fatal exception: null
java.lang.RuntimeException: An error was encounted. Terminating BEAST
	at beast.app.util.ErrorLogHandler.publish(Unknown Source)
	at java.util.logging.Logger.log(Logger.java:738)
	at java.util.logging.Logger.doLog(Logger.java:765)
	at java.util.logging.Logger.log(Logger.java:788)
	at java.util.logging.Logger.severe(Logger.java:1464)
	at beast.app.beastapp.BeastMain.<init>(Unknown Source)
	at beast.app.beastapp.BeastMain.main(Unknown Source)
	at beast.app.beastapp.BeastLauncher.main(Unknown Source)
```

**Comment by Anaphory** [comment]
Looks like there's a beast issue with `useAmbiguities="true"` in `AncestralStateLikelihood`.

**Comment by Anaphory** [comment]
By which I mean, if I set `useAmbiguities="false"`, the two substitutions above seem sufficient, and looking at the TL and ASTL code suggest that TreeLikelihood expects ambiguous tips to need Partials, and doesn't set States, but ASTL seems to only rely on States. (No idea how to fix it, for now I just decided on ignoring ambiguities.)

**Comment by Anaphory** [comment]
Implicit recipe above also needs the ASTL added to the TWTL.

**Comment by lmaurits** [comment]
Glad to see you working on this, I've thought for a while now that it should be the next big feature we add.  I want to make use of this in some analyses I'm going to start running in Feb once some data I'm waiting on comes in, so I will probably jump in on this fairly soon.

**Comment by Anaphory** [comment]
Yep, I think it's quite easy from our starting point, in particular if the beast bug gets fixed (I need to report it first, but I think I want to construct a more minimal example for it which will help fixing it).

What syntax do we want to use for it?
```
[model]
model = bsvs
data = data.csv
features = feature1,feature2,feature3
reconstruct = feature1,feature3
```
strikes me as the obvious one, with `*` as special value – any issues with that?

**Comment by lmaurits** [comment]
That seems fine at first glance, but I'll give it some thought.   Does this sample feature values at every internal node?  Or just the root?  Or at all points with an MRCAPrior?

**Comment by Anaphory** [comment]
Ah, right! When thinking about it earlier, I thought there was more to it than just that, and “Which nodes should it sample?” was precisely it. But I think making *that* accessible might require serious work, and this just samples all nodes (including leafs, which is relevant for leafs with ambiguity)

**Comment by lmaurits** [comment]
there is a `sampling_points` (IIRC) option for `[geography]` sections which controls which nodes to sample locations for.  It accepts a comma-separated list of family names or Glottocodes, just like calibrations, and also `root` for the root of the tree.  I am happy to start with just sampling all nodes for now, but if we want to move toward more user control then keeping a consistent interface for geography and ancestral state seems a good idea (I'm open to discussion on whether this means copying the current geography interface or coming up with a new common interface and deprecating what we have now).

**Comment by lmaurits** [comment]
I'm keen to start using this ASAP with a new project.  Do you already have a vague implementation in the works?  Or can I start hacking on it in a branch off develop?  No point in duplicating effort...

**Comment by Anaphory** [comment]
The changes I made to my Beast XML were:
 - replace `TreeLikelihood` with `AncestralStateTreeLikelihood`
 - replace (if there is no rate variation)/supplement (if there is) `TreeWithTraitLogger` with `TreeWithMetaDataLogger`
 - deactivate `useAmbiguities="false"` in `AncestralStateTreeLikelihood`
 - add idref to the `AncestralStateTreeLikelihood` in `TreeWithTraitLogger`

I have not written any Beastling for it.

**Comment by lmaurits** [comment]
Got it, thanks.

**Comment by lmaurits** [comment]
I seem to have this roughly working in a private branch which I'll merge into develop once we cut 1.3.0.   There are some caveats which appear to be limitations of BEAST or BEAST packages for now: it doesn't work with pruned trees and it doesn't work with ascertainment correction (not 100% sure if this last one is my fault or not).  Aside from that, it appears to work.  One oddity I have noticed is that some features appear to be reconstructed as either "0" or "1", others are reconstructed as either "a" or "b".  I'll need to look into that.  Thanks for showing me the path here.

**Comment by Anaphory** [comment]
Do you have to keep it on a private branch? I would like to use some in the near future, so if you could just push your feature branch here, I can use it, review it, test it, it doesn't interfere with `master` or `develop` or 1.3.0 – it's public, but that's about the only sideeffect I can see.

**Comment by lmaurits** [comment]
Done!

**Comment by Anaphory** [comment]
I took it and started a run.

Did you run a covarion model? Could it be that a/b/0/1 are cold and hot in
0 and 1?

--
Gereon

**Comment by Anaphory** [comment]
I'm pretty sure now that a=0 hot and b=1 hot. Also, I pushed to your branch, have a look.

**Comment by lmaurits** [comment]
I've merged the state_reconstruction branch into develop, and removed the waiting for merge tag.

Since you've opened a separate issue for the enhancement to support reconstruction at limited nodes, it seems to me that all we need to close this issue is some testing and some docs?

**Comment by Anaphory** [comment]
Is this worth keeping open? I think this (undocumented and untested) bit of the ancestral reconstruction API might be a dead end, because it creates immensely huge logfiles that are hard to deal with, so I'm tempted to just put it aside completely.

**Comment by Anaphory** [comment]
Whatever I come up with based on `AncestralState[s]Logger` will still need testing and documentation, of course.

**Comment by lmaurits** [comment]
My intention was to close this when tests and docs were in place.  I actually have some crappy preliminary docs written for this (and also path sampling).

Are the log files really so awful?  I think I used this approach to do some ASR of African Glottolog data (a small number of features) and didn't have huge problems.  Maybe if reconstructing all features in a binarised cognate analysis I can see it getting hairy.  But surely it's better than nothing?

Let me read your recent comments in the other ASR Issue and think about this...

**Comment by Anaphory** [comment]
The log files are Nexus, containing non-standard extended Newick, and I found them (reconstructing *one single* feature across 70-or-so tips) impossible to handle using the existing tools due to being so large – each reconstructed value, even if only one character, has an overhead of about 10 characters due how they are given in the tree, and all tools I found read the entire Nexus file into memory before dealing with it internally.

AncestralStateLogger is TSV (good for stream editing) with one tab added per character reconstructed, and if we modify it to support reconstruction in multiple MRCAs, we can also make the output more compact and more like an alignment string (which is kind of what it is anyway).

**Comment by lmaurits** [comment]
Alright, but supposing there are cases where people do want to log all over the tree, and can handle it (e.g. only 24 languages in Uralic and I have access to a machine with 128GB of RAM), if we abandoned `AncestralStateTreeLikelihood` completely we'd be forced to generate quite unwieldy XML to do this with the `AncestralStateLogger`, right?

**Comment by Anaphory** [comment]
It's not just unwieldy, it's nigh-impossible, I should admit.

The reason is essentially, what would you even do with that type of data? It makes hardly any sense to compare the ancestral reconstructions of “the parent of the parent of this taxon” or of “the right child of the left child of the root”, so I cannot see many use cases where you would not be better off explicitly describing a finite set of MRCAs where you want to sample.

The only use case I can see is implicitly conditioning reconstructions on tree topologies, and running the whole analysis of that using one Beast run instead of several, but then you don't actually have a guarantee that the topologies you care about are sampled.

**Comment by lmaurits** [comment]
What if you want to know where a certain change occurred?

**Comment by Anaphory** [comment]
That makes some sense…

Even with an explicit hypothesis to be tested, that would require sampling an MRCA *and* its parent, which is not a thing `AncestralState[s]Logger` could do at the moment. Hm…

**Comment by lmaurits** [comment]
To be a little less vague, suppose some feature is mostly absent in your dataset but does occur with some non-zero frequency.  You might want to ask "how many times did X evolve?" or "how old was the earliest occurrence of X?".  These both seem like valid research questions, and I can't think of how to answer them without reconstructing at every node.

**Comment by Anaphory** [comment]
I got your meaning. I was saying that even if I weaken that question a lot, my current structure cannot be used to answer it. I think I will put it back in, have a look at #145 for the syntax I would suggest for chosing which of the two options it will be.

**Comment by lmaurits** [comment]
FYI, there is now some (very rough and probably not oudated) documentation of ASR in `docs/advanced.rst`.  This is where proper documentation should go once we've got this finalised.

**Comment by lmaurits** [comment]
From the unhelpful bug reports department: ASR as currently implemented in develop (not sure how out of synch that may be with your own work) does not appear to play nicely with ascertainment correction.  It's not quite as bad as "doesn't work at all", I've seen it work when reconstructing a single feature (the tree logs show it reconstructing values for the dummy columns), but when reconstructing many features the chain just will not start.  For some strange reason it fails to compute a valid starting likelihood for some features, even ones which aren't being reconstructed.  I have no insight into how/why.

**Comment by Anaphory** [comment]
Oh, that's really strange. I'll have to finish some things about our online database first, but as soon as that is mostly done, this is back on my list of things to look at.
Given that I only know the very basic principle of ascertainment correction, I cannot guarantee that I will get anywhere useful, but I'll try.

**Comment by Anaphory** [comment]
- [ ] The current state of full-tree ASR still needs documentation (`reconstruct_at = *`)
 - [ ] We haven't fixed the ascertainment correction problem

**Comment by lmaurits** [comment]
I will try to shed some light on the ascertainment situation this week, can you handle the ASR documentation?  Did you see my comment above about `docs/advanced.rst`?

**Comment by Anaphory** [comment]
I saw that, and that you committed to that file, but I didn't read what you wrote yet. I'm happy to look at it today and adapt it to how I implemented both options.

**Comment by lmaurits** [comment]
Quick notes:

- The problem I was encountering with ascertainment correction happened when using the "old style" `AncestralStateTreeLikelihood`, rather than the nicer `AncestralStateLogger`.  We _are_ still using both of these, for different purposes, right, or have I lost track of things?  If we are not using `ASTL` then possibly my bug is irrelevant.
- On toy examples, `ASL` seems to _work_ with ascertainment correction, in the sense that the analysis actually runs (although I could get even `ASTL` to run sometimes in toy examples), but it _is_ producing slightly ugly output, in that the log file contains columns corresponding to the "dummy sites" used by BEAST to implement ascertainment correction.  Nobody wants to see these.  However, as far as I can see this is a BEAST bug (whoever wrote `ASL` did not consider the possibility of it being used with a `FilteredAlignment` - and to be fair, depending upon the order in which these things were written that may have been causally impossible rather than just careless).  I think this is easy to fix so I might patch BEASTlabs to this end soon.  No work on BEASTling required for 1.4.
- I notice that `ASL` has a `value` input where you can specify a column name for the log file.  It looks like we are not using this and (if I'm not extrapolating incorrectly from my toy examples) when we reconstruct multiple features they are _all_ going to have the same column name "Site0", which is not very user friendly.  Is there a good reason we are not using this to assign nice column names?  Preliminary tests suggest this works fine, although this _does_ break analyses with ascertainment correction because it wants one column name per site, including the dummy sites.

**Comment by Anaphory** [comment]
> - The problem I was encountering with ascertainment correction happened when using the "old style" AncestralStateTreeLikelihood, rather than the nicer AncestralStateLogger. We are still using both of these, for different purposes, right, or have I lost track of things? If we are not using ASTL then possibly my bug is irrelevant.

Yes, we are using both. If someone wants to reconstruct everywhere (`reconstruct_at = *`), they get an ASTL.

> - but it is producing slightly ugly output, in that the log file contains columns corresponding to the "dummy sites" used by BEAST to implement ascertainment correction. Nobody wants to see these.

I can take out the dummy columns, if I know how to recognize them. There are other issues with it, as mentioned before, and I found yet another one today. Follow [my attempts](https://github.com/Anaphory/statereconstruction/blob/master/src/lucl/beast/statereconstruction/AncestralStatesLogger.java) and [this issue to beastlabs](https://github.com/BEAST2-Dev/BEASTLabs/issues/5) for future developments.

> - I notice that ASL has a value input where you can specify a column name for the log file. It looks like we are not using this and (if I'm not extrapolating incorrectly from my toy examples) when we reconstruct multiple features they are all going to have the same column name "Site0"

Yes, I noticed that today, as well. I'll try to fix it now.

**Comment by Anaphory** [comment]
> I can take out the dummy columns, if I know how to recognize them.

Hm, do I just leave out the first column, depending on this boolean?

		ascertained = dataInput.get().isAscertained;

**Comment by lmaurits** [comment]
Not quite, just a second, I am writing up instructions for this.

**Comment by lmaurits** [comment]
>  can take out the dummy columns, if I know how to recognize them.

`Alignment` (I was misremembering when I said `FilteredAlignment` above) has 3 integer inputs relevant to this, `excludefrom`, `excludeto` and `excludeevery`.  If used, they tell BEAST to treat every `excludeevery`-th column, starting from `excludefrom` up to (but not including) `excludeto`, as a dummy column for ascertainment correction.  These are the ones you need to not log.  This is easier than it sounds, because:

- BEASTling never ever uses `excludevery`, so it always has the default value of 1 and hence you can ignore this functionality
- BEASTling always puts its ascertainment dummy columns _before_ the real data, so `excludefrom` is always 0 (which, now that I think of it, is the default value so we could actually not explicitly state this and make our XMLs smaller).

So, basically, you can just look at `excludeto` and this will tell you how many consecutive columns to ignore from the start of the data.  This will be either 1 (if correcting only for binary coding of cognate data) or the size of the statespace (if correcting for non-constancy of data).

**Comment by Anaphory** [comment]
> But given that I modify a Beast class, I should probably give it a `logExcludedColumns` input, defaulting to `false`, which makes the logger skip the excluded columns. I think I can do that.

Strike that. Data has `excludedPatternIndices`, I'll just use that and not give any excluded columns at any time.

**Comment by lmaurits** [comment]
Yes, that sounds good.  I suppose you should do the right thing and try to treat `excludefrom` and `excludeevery` correctly rather than doing a BEASTling-compatible-only job and fixing them to 0 and 1.  Hopefully this is not too hard, it isn't conceptually difficult (the semantics are identical to Python list slicing), just probably grossly verbose in Java.

**Comment by Anaphory** [comment]
I can just check indices in the `data.excludedPatternIndices` set, so yes, it contains the usual levels of Java verbosity, in particular for generating nice headers now (with the convenience of allowing us to to specify classical ASL headers now including columns that will be excluded in the future, and ASsL will take care of removing unnecessary `value` elements when we switch to it in the future).

**Comment by Anaphory** [comment]
> (although I could get even `ASTL` to run sometimes in toy examples)

So the issue is that ASTL does not *run* with ascertainment correction?
In some sense, that is better than creating output which does not actually reflect the model specified.

Can you give an example of a config where that's an issue?

**Comment by lmaurits** [comment]
Yes, I had a large ASTL analysis with ascertainment correction which would not run (could not find an initial state with valid starting likelihood), but was perfectly happy as soon as I turned ascertainment correction off.  However, a stripped down version of the same analysis _did_ run with ascertainment correction on (and reconstructed dummy columns, just like ASL does), so there is something subtle going on.

I'll see if I can find that config and email it to you...
